### ISSUE TITLE
New API: Rewarded video support

### ISSUE THREAD

**Issue by apollonin** [floatinghotpot/cordova-admob-pro#332]
Hi, admob is now supports rewarded video. 
http://www.awesomescreenshot.com/image/1089806/eb73f96955f1226099cca4c58709659b
May you please add that support into this awesome plugin?

**Comment by floatinghotpot** [comment]
I do have the plan, but it takes time. Stay tuned. :-)

**Comment by apollonin** [comment]
I'm glad to hear it!

**Comment by Mrglut** [comment]
Hi. this feature really need! 
Thx for this good project!

**Comment by Mrglut** [comment]
@floatinghotpot Sory, but whe you can add this feature?

**Comment by floatinghotpot** [comment]
Reward video Ad is now supported in plugin version v2.16.0

New APIs:
```javascript
// use reward video
AdMob.prepareRewardVideoAd(adId/options, success, fail);
AdMob.showRewardVideoAd();
```

See [docs](https://github.com/floatinghotpot/cordova-admob-pro/tree/master/docs#admobpreparerewardvideoadadidoptions-success-fail) for details.

You need configure mediation in AdMob portal, and mediate your ad unit id to other ad networks providing reward videos.

To give user reward when he/she watched the video, you need listen to event 'onAdPresent', check the adType 'rewardvideo'.

```javascript
document.addEventListener('onAdPresent', function(e){ 
  if(e.adType == 'rewardvideo') { 
    console.log( e.rewardType ); 
    console.log( e.rewardAmount ); 
  } 
});
```

**Comment by apollonin** [comment]
After updating to cordova-plugin-admobpro 2.16.0 "AdMob Plugin Pro"
I've got strange error while building the project:

```
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:103: error: method does not override or implement a method from a supertype
  @Override
  ^
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:296: error: method does not override or implement a method from a supertype
  @Override
  ^
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:319: error: method does not override or implement a method from a supertype
  @Override
  ^
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:635: error: cannot find symbol
      fireAdErrorEvent(EVENT_AD_FAILLOAD, errorCode, getErrorReason(errorCode), ADTYPE_REWARDVIDEO);
                                                                                ^
  symbol:   variable ADTYPE_REWARDVIDEO
  location: class AdMobPlugin.RewardVideoListener
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:640: error: cannot find symbol
      fireAdEvent(EVENT_AD_LEAVEAPP, ADTYPE_REWARDVIDEO);
                                     ^
  symbol:   variable ADTYPE_REWARDVIDEO
  location: class AdMobPlugin.RewardVideoListener
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:648: error: cannot find symbol
      fireAdEvent(EVENT_AD_LOADED, ADTYPE_REWARDVIDEO);
                                   ^
  symbol:   variable ADTYPE_REWARDVIDEO
  location: class AdMobPlugin.RewardVideoListener
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:650: error: cannot find symbol
      if(autoShowRewardVideo) {
         ^
  symbol:   variable autoShowRewardVideo
  location: class AdMobPlugin.RewardVideoListener
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:651: error: cannot find symbol
        showRewardVideoAd();
        ^
  symbol:   method showRewardVideoAd()
  location: class AdMobPlugin.RewardVideoListener
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:657: error: cannot find symbol
      fireAdEvent(EVENT_AD_WILLPRESENT, ADTYPE_REWARDVIDEO);
                                        ^
  symbol:   variable ADTYPE_REWARDVIDEO
  location: class AdMobPlugin.RewardVideoListener
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:662: error: cannot find symbol
      fireAdEvent(EVENT_AD_WILLPRESENT, ADTYPE_REWARDVIDEO);
                                        ^
  symbol:   variable ADTYPE_REWARDVIDEO
  location: class AdMobPlugin.RewardVideoListener
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:667: error: cannot find symbol
      fireAdEvent(EVENT_AD_DISMISS, ADTYPE_REWARDVIDEO);
                                    ^
  symbol:   variable ADTYPE_REWARDVIDEO
  location: class AdMobPlugin.RewardVideoListener
/var/www/sites/public/app/platforms/android/src/com/rjfun/cordova/admob/AdMobPlugin.java:675: error: cannot find symbol
              obj, ADTYPE_REWARDVIDEO, EVENT_AD_PRESENT, reward.getType(), reward.getAmount());
                   ^
  symbol:   variable ADTYPE_REWARDVIDEO
  location: class AdMobPlugin.RewardVideoListener

```


There was not any error in previous version...

**Comment by floatinghotpot** [comment]
Please also update dependent plugin `cordova-plugin-extension`

**Comment by apollonin** [comment]
Hm... that helps. 

Do i need to add some other plugins for rewarded video mediators? For example some SDKs  of adColony, Applovin, etc? 
Or admob pro plugin is the only one required for that? 

Thanks for your response!

**Comment by floatinghotpot** [comment]
Yes, some other SDKs is needed. It depends on what Ad networks you configured for mediation in AdMob portal.

**Comment by apollonin** [comment]
oh, and did you test some of them, when developed this admob plugin? Could you share some experience or ref links please? 
I don't see the way to integrate adcolony (for example), 'cos they don't have cordova plugin...

**Comment by floatinghotpot** [comment]
Here are the AdMob mediation networks:
* For Android: https://developers.google.com/admob/android/mediation-networks
* For iOS: https://developers.google.com/admob/ios/mediation-networks#supported-ad-networks

For easier to use, the SDKs and adapters may need be packed into cordova plugins. With some priority, which one do you guys prefer to use for reward video?

**Comment by apollonin** [comment]
yeah, packing the, into cordova plugins would be a great idea!
as i've researched here is some kind of priority list of rewarded video mediators: 

AdColony
Applovin
Chartboost
Fuse
Unity

**Comment by apollonin** [comment]
May i expect you to make (at least adColony example)  plugin? So I can see how it is done and maybe help you to make plugins for another networks?

**Comment by floatinghotpot** [comment]
I will have a try, but it takes time, so stay tuned.

**Comment by apollonin** [comment]
i'm here ;)

**Comment by NathanVss** [comment]
I wanted to test a rewarded video ad with the new API. 
I setup the admob mediation with AdColony, and it doesn't work with this plugin. Is that normal ? 
Thanks

**Comment by BadSector82** [comment]
@NathanVss I have successfully implemented AdColony Rewarded Video mediation using this plugin. 

There are a couple of important things:

1. Add the AdColony initialization calls to your main Activity in the onCreate, onResume and onPause methods (Step1 & 2 here https://github.com/AdColony/AdColony-Android-SDK/wiki/Showing-V4VC-Videos)
   If you haven't done this, you will see this message:

   `E/AdColony: AdColony methods used before initialization, Activity reference is null. Disabling AdColony until next launch.`

2. Make sure to copy adcolony.jar to your platform's "libs" folder (download Admob compatible version here: https://developers.google.com/admob/android/mediation-networks#supported-ad-networks)
3. For rewarded videos, remember to switch your AdColony's Zone type from "Preroll/Interstitial" to "Value Exchange/V4VC"

After these steps, I was seeing an error from this plugin's AdMobPlugin.java:
 
`java.util.UnknownFormatConversionException: Conversion: i`
 
I have created a pull request to fix this https://github.com/floatinghotpot/cordova-admob-pro/pull/367

**Comment by WonderPolls** [comment]
Would love to have a Vungle plugin for this.  Thanks!

**Comment by crownclown90** [comment]
im trying to impliment Rewarded videos but i am not Not receiving the onAdPresent event or the onAdDismiss events

as a result i am unable to rewards players without this event  


Help?

**Comment by prantikv** [comment]
If I understand this correctly and correct me if I am wrong.
To show AdMob reward video ads a mediation network is necessary otherwise no ads will show up. Is that true?

Does AdMob have an option to setup ads what show from the AdMob network only, just live banner ads?

**Comment by EvezEliyev** [comment]
How can I do that when the video ad is not visible x, so you can not close the ad?
How can I do it?
![image](https://user-images.githubusercontent.com/19881218/29799614-f35ac0c4-8c75-11e7-924b-94bcdc57da56.png)

**Comment by floatinghotpot** [comment]
Is the number at right top corner a count down? 

I just guess, if it's a count down, google just design it by intention, after a timeout, then allow you close it.

**Comment by ranjhanisaddam** [comment]
First of all I want appreciate for your efforts that we become able to earn from admob through your app for hybird apps. (Claps)
Its been few days I have implemented Reward video, Tests went excellent, I have released update for my app and then tested the update in friends mobile they are being presented with reward video which is ending successfully and users are getting their due reward.
Problem is I am not getting any impression or earning on admob dashboard it displays $0.00 for video reward ad unit.

Does this seems familiar to you?
